### PR TITLE
perf(dashboard): Improve performance of complex dashboards

### PR DESCRIPTION
**Pull request by kgabryje**
### SUMMARY
This PR improves the performance of dashboards with complex structure (multiple tabs, many levels of nesting, many charts, native filters).
The main culprit of poor performance was a function used for finding charts in filter's scope. `getChartIdsInFilterScope` was written in a very suboptimal way, which drastically slowed down loading charts - which was especially noticeable when opening a dashboard or switching tabs.
I've rewritten that function and made a few other improvements, which resulted in a much smoother experience.
The scope of the changes in this PR is related only to native filters and cross filters. Performance issues related to filter box were not addressed, as the filter box will soon be deprecated in favor of native filters.

The tests were performed on a dashboard with 100+ charts, 70+ tabs, up to 4 levels of nested tabs, up to 10 charts in a single tab, 9 native filters.

Dashboard used for testing: [dashboard_export_20220309T112451.zip](https://github.com/apache/superset/files/8213872/dashboard_export_20220309T112451.zip)

### BEFORE/AFTER SCREENSHOTS OR ANIMATED GIF
Before:

https://user-images.githubusercontent.com/15073128/157247412-84be2244-b26c-48b4-9eb4-4005ffc4ebcd.mov

After:

https://user-images.githubusercontent.com/15073128/157247602-359b5bab-e857-466f-a6d4-dfc8422fcd48.mov

### TESTING INSTRUCTIONS
1. Verify that building dashboards works like before
2. Verify that applying native filters works like before
3. Verify that applying cross filters works like before
4. Verify that applying filters from filter box works like before

### ADDITIONAL INFORMATION

- [ ] Has associated issue:
- [ ] Required feature flags:
- [ ] Changes UI
- [ ] Includes DB Migration (follow approval process in [SIP-59](https://github.com/apache/superset/issues/13351))
  - [ ] Migration is atomic, supports rollback & is backwards-compatible
  - [ ] Confirm DB migration upgrade and downgrade tested
  - [ ] Runtime estimates and downtime expectations provided
- [ ] Introduces new feature or API
- [ ] Removes existing feature or API

https://app.shortcut.com/preset/story/38683/performance-issues-to-navigate-between-tabs-depending-on-the-amount-of-charts-nested-tabs-on-a-dashboard

### CODE REVIEW

**Comment by rusackas** [comment]
@graceguo-supercat is going to LOVE this PR 🚀 

@kgabryje could you please export your test dashboard (if it uses sample data) and attach it here* and/or import it into the above ephemeral environment when it spins up?

_*Fragility caveat that the import might get nuked with the ephemeral env, thus the request to upload/link here ;)_

**Review by villebro — APPROVED** [review]
This is incredible work, a major achievement in improving dashboard performance! 👏  Changes look very straight forward and would actually serve as a great example for frontend development guidelines (use custom hooks to encapsulate complex state logic, memoize heavy computation and in the process avoid the need for `JSON.stringify` when referencing the results in dependency arrays etc). Also, not a blocker, but it would be great if `getChartIdsInFilterScope` could be converted to TS.

**Comment by jinghua-qa** [comment]
> @graceguo-supercat is going to LOVE this PR 🚀
> 
> @kgabryje could you please export your test dashboard (if it uses sample data) and attach it here* and/or import it into the above ephemeral environment when it spins up?
> 
> _*Fragility caveat that the import might get nuked with the ephemeral env, thus the request to upload/link here ;)_

Would be helpful if the perf test dashboard can be upload, currently the import dashboard feature seems not enable in the ephemeral env and i can not import the test dashboard

**Comment by graceguo-supercat** [comment]
@kgabryje Thank you for the perf improvement! just 1 question here: I saw both `getChartIdsInFilterBoxScope` and `getChartIdsInFilterScope` functions. So when to use each?

File: superset-frontend/src/dashboard/util/activeDashboardFilters.js
Comment (kgabryje) [reply superset-frontend/src/dashboard/util/activeDashboardFilters.js]: Done, good point!

File: superset-frontend/src/dashboard/components/DashboardBuilder/DashboardContainer.tsx
Comment (kgabryje) [reply superset-frontend/src/dashboard/components/DashboardBuilder/DashboardContainer.tsx]: 👍

**Comment by kgabryje** [comment]
> @kgabryje Thank you for the perf improvement! just 1 question here: I saw both `getChartIdsInFilterBoxScope` and `getChartIdsInFilterScope` functions. So when to use each of them?

@graceguo-supercat I kept `getChartIdsInFilterBoxScope` because refactoring the logic related to filter box would be risky as that code is quite complex - I decided not to do it since we're on the road of deprecating filter box in favour in native filters. I added a comment to `getChartIdsInFilterBoxScope` that redirects to the new `getChartIdsInFilterScope`

**Comment by kgabryje** [comment]
@rusackas @jinghua-qa I added exported dashboard to the PR description 🙂 I'll spin up a new test env with `VERSIONED_EXPORT` enabled once docker builds

File: superset-frontend/src/dashboard/components/FiltersBadge/selectors.ts
Comment (michael-s-molina) [inline superset-frontend/src/dashboard/components/FiltersBadge/selectors.ts:34]:
```suggestion
import { CHART_TYPE } from 'src/dashboard/util/componentTypes';
```

File: superset-frontend/src/dashboard/components/FiltersBadge/selectors.ts
Comment (michael-s-molina) [inline superset-frontend/src/dashboard/components/FiltersBadge/selectors.ts:33]:
```suggestion
import { Layout } from 'src/dashboard/types';
```

**Comment by kgabryje** [comment]
/testenv up FEATURE_VERSIONED_EXPORT=true FEATURE_DASHBOARD_CROSS_FILTERS=true

**Review by michael-s-molina — APPROVED** [review]
LGTM. Thanks for the awesome work @kgabryje. I left some non-blocking comments. 

Given the critical nature of the components, I kindly ask that you wait for @jinghua-qa's approval before merging.

File: superset-frontend/src/dashboard/components/DashboardBuilder/DashboardContainer.tsx
Comment (kgabryje) [reply superset-frontend/src/dashboard/components/DashboardBuilder/DashboardContainer.tsx]: Good idea! Done

**Review by rusackas — APPROVED** [review]
LGTM, and feels super snappy in testing. This PR is such a huge improvement!!

**Review by jinghua-qa — APPROVED** [review]
LGTM

**Comment by sadpandajoe** [comment]
🏷️  preset:2022.9

**Pull request merged**
